Add check for keywords in PR titles to block merge

corbin-coleman opened the pull request:
Add another acceptable configuration keyword, `merge_block_keywords` that can be added to the `.unir.yml` file. If the PR title contains any of the words in the `merge_block_keywords` section then the PR will not be able to automatically be able to merge.

A good example of usage would be with `WIP` PRs. They shouldn't be merged until they are no longer a `WIP`, but they may get an appropriate number of `LGTM`s before that happens.

The keywords are defaulted to `WIP:` if no keywords are defined in the `.unir.yml` file.

Signed-off-by: corbin-coleman <corbin.coleman@docker.com>

corbin-coleman committed:
Add check for keywords in PR titles to block merge

Add another acceptable configuration keyword, `merge_block_keywords` that can be added to the `.unir.yml` file. If the PR title contains any of the words in the `merge_block_keywords` section then the PR will not be able to automatically be able to merge.

A good example of usage would be with `WIP` PRs. They shouldn't be merged until they are no longer a `WIP`, but they may get an appropriate number of `LGTM`s before that happens.

The keywords are defaulted to `WIP:` if no keywords are defined in the `.unir.yml` file.

Signed-off-by: corbin-coleman <corbin.coleman@docker.com>

corbin-coleman committed:
Use strings.Contains

Signed-off-by: corbin-coleman <corbin.coleman@docker.com>

jose-bigio commented:
Here's the build error

```
internal/webhook.go:252:5: undefined: reviewEvent
internal/webhook.go:269:4: too many arguments to return
	have (bool)
	want ()
internal/webhook.go:272:2: too many arguments to return
	have (bool)
	want ()
internal/webhook.go:326:28: checkMergeBlockKeywords(config.MergeBlockKeywords, prTitle) used as value
make: *** [build/unir] Error 2
```

jose-bigio commented on internal/webhook.go:
```diff
@@ -249,17 +249,30 @@ func handleStatus(integrationID int, keyfile string, statusEvent github.StatusEv
 			*statusEvent.Repo.Name,
 			*pullRequest.Head.SHA,
 			*pullRequest.Number,
+			*reviewEvent.PullRequest.Title,
```
I'm guess this should be 
*pullRequest.Title

jose-bigio commented on internal/webhook.go:
```diff
@@ -249,17 +249,30 @@ func handleStatus(integrationID int, keyfile string, statusEvent github.StatusEv
 }
 
-func mergePullRequest(client *github.Client, owner, repo, sha string, prNumber int) {
+func checkMergeBlockKeywords(mergeBlockKeywords []string, prTitle string) {
```
The function signature doesn't indicate that a bool is being returned (which is where the other errors are coming from).

corbin-coleman committed:
Make changes so it builds

Thanks to Jose for figuring out the necessary changes

s/reviewEvent.PullRequest.Title/pullRequest.Title/
checkMergedBlockKeywords needed to specify it returned a bool

Signed-off-by: corbin-coleman <corbin.coleman@docker.com>

seemethere commented on internal/config.go:
```diff
@@ -5,10 +5,11 @@ import (
+	ApprovalsNeeded    int      `yaml:"approvals_needed"`
+	ConsensusNeeded    bool     `yaml:"consensus_needed"`
+	MergeMethod        string   `yaml:"merge_method"`
+	MergeBlockKeywords []string `yaml:"merge_block_keywords"`
```
~Can we also add a test for this?~

Just to make sure we can actually read these in.

corbin-coleman committed:
Add test to make sure the new bit works

Signed-off-by: corbin-coleman <corbin.coleman@docker.com>

jose-bigio commented:
To deploy we will need to create a local webhook handler. 
This handler will be something similar to: 
https://github.com/seemethere/unir/blob/master/internal/webhook.go#L23-#L32

In particular we will want to implement something like this: 
https://github.com/docker/batterup/blob/master/internal/webhook.go#L17-#L26 

Note that for the above code block an api token is utilized as opposed to an integration id. 
This will enable us to test `unir` locally without having to worry about setting up an app.  In order to toggle between these routers will need to add a flag in the cli to specify whether  we want the app based router or the webhook based router.

jose-bigio commented:
Partial illustration of this working: 

https://github.com/jose-bigio/Webhooks/pull/28

Commit:
Modify code to allow for testing option

Signed-off-by: jose-bigio <jose.bigio@docker.com>

golangci commented on internal/webhook.go:
```diff
@@ -184,20 +200,30 @@ func GrabConfig(ctx context.Context, client *github.Client, repo, owner string,
+			&oauth2.Token{AccessToken: apiToken},
+		)
+		return github.NewClient(oauth2.NewClient(ctx, ts))
+	} else {
```
if block ends with a return statement, so drop this else and outdent its block

golangci commented on cmd/unir/main.go:
```diff
@@ -18,22 +19,40 @@ var (
+
+	// Use either a test handler which allows you to use a webhook
+	// or the regular handler which is integrated with an app
+	router := mux.NewRouter()
```
ineffectual assignment to `router`

seemethere replied:
I agree with @golangci here, this doesn't need to be assigned here

seemethere commented on cmd/unir/main.go:
```diff
@@ -18,22 +19,40 @@ var (
+	id := (os.Getenv(INTEGRATION_ID))
+
 	debug := flag.Bool("debug", false, "Toggle debug mode")
+	test := flag.Bool("test", false, "Toggle for testing")
```
We should just rename this to `oauth` mode

seemethere commented on internal/webhook.go:
```diff
@@ -310,6 +360,11 @@ func mergePullRequest(client *github.Client, owner, repo, sha string, prNumber i
 	}
 
+	if checkMergeBlockKeywords(config.MergeBlockKeywords, prTitle) {
+		doStatus("checking keywords that block merging", "failure", "Can't merge currently, title contains keywords that block merges.")
```
I think the commit context should read:

> Automatic merging blocked, title contains keywords that prevent unir from automatically merging

seemethere commented on internal/webhook.go:
```diff
@@ -12,19 +12,35 @@ import (
 		Secret:        secret,
 		integrationID: integrationID,
 		keyfile:       keyfile,
+		testClient:    false,
```
Should be renamed to `needsOAuthClient`, since that's literally what it does.

Commit:
Make style changes to Pull Request

Signed-off-by: jose-bigio <jose.bigio@docker.com>

Commit:
Update vendoring

Signed-off-by: jose-bigio <jose.bigio@docker.com>

Commit:
Update commit context

Signed-off-by: jose-bigio <jose.bigio@docker.com>

jose-bigio commented:
It's green!

seemethere reviewed (approved):
LGTM
